### ISSUE TITLE
tlogclient should limit the amount of times it tries to reconnect to

### ISSUE THREAD

**Issue by GlenDC** [zero-os/0-Disk#541]
Right now the tlogclient can get stuck in an infinite loop if during a reconnect no tlogserver is available, and thus it will keep shifting and trying without success.

You can see it clearly for yourself how the infinite loop can happen (and has happened for some users): https://github.com/zero-os/0-Disk/blob/a60cf2b832cfa94ceafd4f0d2fec052cd47184fa/tlog/tlogclient/client.go#L323-L342

+ Even though the return parameter is `error`, and even though that result is handled by the callee, the function will in fact never return an error, and instead keep trying until it can return `nil`.
+ We also receive an input parameter which indicates the point in time at which we started retrying (as the callee uses `time.Now()`), this input parameter is not used;

There needs to be put some limit here:

+ Either we limit based on some maximum time of trying to reconnect;
+ Or we limit based on how many times we can retry each server or the total amount of retries;

But right now we simply keep trying, which is pointless, as in that scenario all hope is basically lost for the vdisk and we shouldn't block the entire system because of it, and instead exit with an error.

**Comment by GlenDC** [comment]
It might be a bit more invasive than what you had planned to do @chrisvdg, but what do you think about the following approach:

```go
// reconnect to server
func (c *Client) reconnect(closedTime time.Time) error {
	defer c.resetServerUsedProperty()
	var err error

	log.Info("tlogclient reconnect")
	for {
		// try other server
		err = c.shiftServer()
		if err != nil {
			return err
		}

		// TODO: obviously you would also have to change the c.connect function
		// such that it gets the address from the new server structure
		if err = c.connect(); err == nil {
			// if reconnect success, sent all unflushed blocks
			// because we don't know what happens in servers.
			// it might crashed
			c.blockBuffer.SetResendAll()
			return nil
		}
	}
}

// shift server address move current active client to the back
// and use 2nd entry as main server
func (c *Client) shiftServer() error {
	c.mux.Lock()
	defer c.mux.Unlock()

	for i := range c.servers {
		if !c.servers[i].used {
			c.servers[i].used = true
			c.servers = append(c.servers[i+1:], c.servers[:i])
			return nil
		}
	}

	return errors.New("can't shift as all servers have been exchausted")
}

func (c *Client) resetServerUsedProperty() {
	c.mux.Lock()
	defer c.mux.Unlock()
	for i := range c.servers {
		c.servers[i].used = false
	}
}

// ...
// this last part is more pseudo-code, just to show you how simple the structure for the internal in-memory server type could stay
c.servers = []struct{
	address string
	used bool
}
// ...
```

**Comment by chrisvdg** [comment]
Indeed a bit more invasive than I imagined this issue would be but this would allow to retry each server 1 time.
But why not for loop for the `len(c.servers)` then as order is guaranteed in slices, if it were a map we would indeed need to do something like this, no? Or does this also deal with the issue of when the servers get updated while retrying?

**Comment by GlenDC** [comment]
I don't think get it. We for loop the `c.server`, and the order is guaranteed as we are using a slice. Where do you see a map?

This approach would also deal with the issue of when the servers get updated yes, as all servers would be new, as you create/overwrite this in-memory body (again) when you re-load servers, and thus the used property would be false, once again.

This might mean that some servers which we retried before in the past get retried again (as in an extra time (or two)). This is OK, as you can assume that in most cases a config gets reloaded because AYS fixed some broken servers, or something like that.

**Comment by chrisvdg** [comment]
I meant to for loop for each server in `reconnect`

Do we need that though? Looking at `run` where `reconnect` is called, will loop till it can be reconnected only with a log in between reconnect attempts. So the next time it will try to reconnect it will be with the updated servers anyways?

**Comment by GlenDC** [comment]
Yes you're right. While the loop does work functionally, you are indeed correct in stating that it is not needed. But you know, I was more trying to give code to help me describe a possible approach. Don't get too hang up on the details of the code.

You could replace:

```go
for i := range c.servers {
    if !c.servers[i].used {
        c.servers[i].used = true
        c.servers = append(c.servers[i+1:], c.servers[:i]...)
        return nil
    }
}

return errors.New("can't shift as all servers have been exchausted")
```

with:

```go
c.servers[0].failed = true
c.servers = append(c.servers[1:], c.servers[0])
if c.servers[0].failed {
    return errors.New("all servers have been exchausted") 
}
return nil
```

As you're right, the order is static, so there is no need to go through a loop. As in, we can assume that a soon as we saw a server that we already shifted before, we could simply assume that we finished the list.

In this change I also use a failed property, instead of a used property, as I think it makes more sense. Because only when we shift a server back to the end, we know for sure that it has failed (been used) for sure, and thus only right before we do that we should update that property.

With that being said. You can even improve the design more. We could for example add a timestamp property which indicates when the server was last tried.

In that way you could also change the logic proposed above, from:

```go
c.servers[0].failed = true
c.servers = append(c.servers[1:], c.servers[0])
if c.servers[0].failed {
    return errors.New("all servers have been exchausted") 
}
```

to:

```go
c.servers[0].failed = true
c.servers[0].shiftTime = time.Now()
c.servers = append(c.servers[1:], c.servers[0])
if c.servers[0].failed {
    if (time.Now().Subtract(c.servers[0].shiftTime) > someConstantDurationLimit {
      return nil // retry again anyhow, as it has been failed a long time ago...
    } 
    return errors.New("all servers have been exchausted") 
}
```

**Comment by iwanbk** [comment]
After some discussion with @GlenDC on telegram, I think (and assumed @GlenDC agree) that this issue is not suitable as `mentor` type, the solution still not clear and i think we might need to change the API.

For @chrisvdg , i feel sorry about this, but this work has started while i'm on vacation so i couldn't do anything at that time.
There are some obvious things which can be done by @chrisvdg  here:
- report to AYS
- give some delay between reconnect
- limit the logging 

After that, i will assign this issue to me.

FYI, Some parts of PR #548 that i don't like:
- it doesn't retry the reconnect, doesn't handle network glitch
- the change is too invasive (lock, channel that blocks, among other things) for such rare case (assumed that self healing do the work properly), but yeah we might need this in the end.
- in case of a block failed to be sent, the sequence number in the nbd cache is not being reordered. It means that subsequent blocks will have wrong sequence

**Comment by zaibon** [comment]
Retrying all the servers once and then propagate the error to AYS with STDERR sounds a reasonable solution. I also like that is mimic the same behavior as the one used for ardb connection as said by @GlenDC  
>having it just only tries once, does fit our policy that we apply also in the ARDB side, where we also just try a server once, before just giving up on it until further notice from AYS.

**Comment by iwanbk** [comment]
I can see two big issues here, CMIIW

(1) `tlogclient.Send` will block in case of all servers failed

one possible solution is to simply return error as already done by @chrisvdg at #548, but it is still not enough, we also need to do these things:
- resend the block after the client become connected again. So we don't need to reorder the sequence number, and tlog stay consistent with nbd
- set some flag to make nbd's `SetBlock` failed. It is to prevent our buffer become full. It is the same method as my PR at #613 when the ardb failed.  

We also need the flag in tlogclient

As improvement, we can also only set the flag after the buffer is really full.

(2) current reconnection is very naive

It is something that definitely need to be improved, as pointed in my previous comment and in this issue description

> Retrying all the servers once and then propagate the error to AYS with STDERR sounds a reasonable solution. I also like that is mimic the same behavior as the one used for ardb connection as said by 

While it is true that simply reconnecting forever is stupid, i think we should not stick too hard with this  principle because:
- reconnecting is cheap
- sticking to 1 reconnect policy too hard might make the code much more complex, as what we have in #548 

So, i think we should find the balance between not so stupid and not so complex code.
